sessiontxn/legacy/provider.go

```diff
@@ -72,7 +72,7 @@ func (p *SimpleTxnContextProvider) OnInitialize(ctx context.Context, tp sessiont
 	switch tp {
 	case sessiontxn.EnterNewTxnDefault, sessiontxn.EnterNewTxnWithBeginStmt:
 		txnCtx := sessVars.TxnCtx
-		if tp != sessiontxn.EnterNewTxnWithBeginStmt || txnCtx.History != nil || txnCtx.IsStaleness {
+		if tp != sessiontxn.EnterNewTxnWithBeginStmt || txnCtx.History != nil || txnCtx.IsStaleness || sessVars.SnapshotTS > 0 {
```

Comment:
Better to add some comments here as it's a bit complex by now.

Reply:
done